File: OpenSuperWhisper.App/AutoStart.cs
using Microsoft.Win32;
using OpenSuperWhisper.Core;

namespace OpenSuperWhisper.App;

/// <summary>
/// Toggles "start with Windows" via the per-user Run registry key (HKCU, not HKLM - never
/// needs admin rights and only affects the current Windows account). This only ever runs when
/// the user explicitly checks/unchecks the box in Settings; it never runs on its own.
/// </summary>
public static class AutoStart
{
    private const string RunKeyPath = @"Software\Microsoft\Windows\CurrentVersion\Run";
    private const string ValueName = "OpenSuperWhisper";

    public static bool IsEnabled()
    {
        try
        {
            using var key = Registry.CurrentUser.OpenSubKey(RunKeyPath, writable: false);
            return key?.GetValue(ValueName) is string;
        }
        catch (Exception ex)
        {
            Log.Error("读取开机自启注册表项失败", ex);
            return false;
        }
    }

    public static void SetEnabled(bool enabled)
    {
        try
        {
            using var key = Registry.CurrentUser.OpenSubKey(RunKeyPath, writable: true)
                             ?? Registry.CurrentUser.CreateSubKey(RunKeyPath);
            if (enabled)
            {
                var exePath = Environment.ProcessPath;
                if (string.IsNullOrEmpty(exePath))
                {
                    Log.Error("设置开机自启失败：拿不到当前 exe 路径");
                    return;
                }
                key.SetValue(ValueName, $"\"{exePath}\"");
            }
            else
            {
                key.DeleteValue(ValueName, throwOnMissingValue: false);
            }
            Log.Info($"开机自启已{(enabled ? "开启" : "关闭")}");
        }
        catch (Exception ex)
        {
            Log.Error("设置开机自启注册表项失败", ex);
        }
    }
}

File: src/autostart.rs
//! Launch-at-sign-in toggle via the HKCU `Run` registry key -- mirrors the
//! shipping C# app's `AutoStart.cs`. The registry key itself is the single
//! source of truth (no separate settings file that could fall out of sync
//! with it): `is_enabled()` always reflects the real current state, read
//! fresh, never cached.

use winreg::RegKey;
use winreg::enums::{HKEY_CURRENT_USER, KEY_READ, KEY_WRITE};

const RUN_KEY_PATH: &str = r"Software\Microsoft\Windows\CurrentVersion\Run";
const VALUE_NAME: &str = "Aevocis";

fn open() -> std::io::Result<RegKey> {
    RegKey::predef(HKEY_CURRENT_USER).open_subkey_with_flags(RUN_KEY_PATH, KEY_READ | KEY_WRITE)
}

/// Whether Aevocis is currently registered to launch at sign-in.
pub fn is_enabled() -> bool {
    open().and_then(|key| key.get_value::<String, _>(VALUE_NAME)).is_ok()
}

/// Enables or disables launch-at-sign-in. Returns the underlying registry
/// error (e.g. an unwritable Run key) rather than swallowing it, so the tray
/// menu can tell the user the toggle didn't actually take effect instead of
/// silently showing a checkbox state that lies about reality.
pub fn set_enabled(enabled: bool) -> std::io::Result<()> {
    let key = open()?;
    if enabled {
        let exe = std::env::current_exe()?;
        key.set_value(VALUE_NAME, &format!("\"{}\"", exe.display()))
    } else {
        match key.delete_value(VALUE_NAME) {
            Ok(()) => Ok(()),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(e) => Err(e),
        }
    }
}
